core/src/banking_stage/packet_receiver.rs

```diff
@@ -0,0 +1,169 @@
+    id: u32,
+    packet_deserializer: PacketDeserializer,
+    last_receive_time: Instant,
+}
+
+impl PacketReceiver {
+    pub fn new(id: u32, receiver: BankingPacketReceiver) -> Self {
+        Self {
+            id,
+            packet_deserializer: PacketDeserializer::new(receiver),
+            last_receive_time: Instant::now(),
+        }
+    }
+
+    pub fn do_packet_receiving_and_buffering(
+        &mut self,
+        unprocessed_transaction_storage: &mut UnprocessedTransactionStorage,
+        banking_stage_stats: &mut BankingStageStats,
+        tracer_packet_stats: &mut TracerPacketStats,
+        slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+    ) -> Result<(), RecvTimeoutError> {
+        // Gossip thread will almost always not wait because the transaction storage will most likely not be empty
+        let recv_timeout = if !unprocessed_transaction_storage.is_empty() {
+            // If there are buffered packets, run the equivalent of try_recv to try reading more
+            // packets. This prevents starving BankingStage::consume_buffered_packets due to
+            // buffered_packet_batches containing transactions that exceed the cost model for
+            // the current bank.
+            Duration::from_millis(0)
+        } else {
+            // Default wait time
+            Duration::from_millis(100)
+        };
+
+        let (res, receive_and_buffer_packets_time) = measure!(self.receive_and_buffer_packets(
+            recv_timeout,
+            unprocessed_transaction_storage,
+            banking_stage_stats,
+            tracer_packet_stats,
+            slot_metrics_tracker,
+        ));
+        slot_metrics_tracker
+            .increment_receive_and_buffer_packets_us(receive_and_buffer_packets_time.as_us());
+
+        res
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    /// Receive incoming packets, push into unprocessed buffer with packet indexes
+    fn receive_and_buffer_packets(
+        &mut self,
+        recv_timeout: Duration,
+        unprocessed_transaction_storage: &mut UnprocessedTransactionStorage,
+        banking_stage_stats: &mut BankingStageStats,
+        tracer_packet_stats: &mut TracerPacketStats,
+        slot_metrics_tracker: &mut LeaderSlotMetricsTracker,
+    ) -> Result<(), RecvTimeoutError> {
+        let mut recv_time = Measure::start("receive_and_buffer_packets_recv");
+        let ReceivePacketResults {
+            deserialized_packets,
+            new_tracer_stats_option,
+            passed_sigverify_count,
+            failed_sigverify_count,
+        } = self.packet_deserializer.handle_received_packets(
+            recv_timeout,
+            unprocessed_transaction_storage.max_receive_size(),
+        )?;
+        let packet_count = deserialized_packets.len();
+        debug!(
+            "@{:?} process start stalled for: {:?}ms txs: {} id: {}",
+            timestamp(),
+            duration_as_ms(&self.last_receive_time.elapsed()),
```

Comment:
seems `self.last_receive_time` is only maintained for this `debug!` logging. i think we can remove this altogether.

that further means, maybe we can move these code into `PacketDeserializer`. yeah, i know there's cognitize impedence mismatch, but i thhink `PacketDeserializer` is small still and it's already taking `receiver`. so, i prefer dozen of right-sized `struct`s, instead of smaller `structs`, imo